fuzzers/libafl-fuzz/src/executor.rs

```diff
@@ -0,0 +1,185 @@
+        ));
+    }
+    #[cfg(target_os = "macos")]
+    {
+        // check if the binary is an ELF file
+        if mmap[0..4] != [0x7f, 0x45, 0x4c, 0x46] {
+            return Err(Error::illegal_argument(format!(
+                "Program '{}' is not an ELF binary",
+                bin_path.display()
+            )));
+        }
+
+        #[cfg(not(target_arch = "arm"))]
+        {
+            if (mmap[0] != 0xCF || mmap[1] != 0xFA || mmap[2] != 0xED)
+                && (mmap[0] != 0xCA || mmap[1] != 0xFE || mmap[2] != 0xBA)
+            {
+                return Err(Error::illegal_argument(format!(
+                    "Program '{}' is not a 64-bit or universal Mach-O binary",
+                    bin_path.display()
+                )));
+            }
+        }
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    let check_instrumentation = !opt.qemu_mode
+        && !opt.frida_mode
+        && !opt.unicorn_mode
+        && !opt.forkserver_cs
+        && !opt.non_instrumented_mode;
+
+    #[cfg(target_os = "linux")]
+    let check_instrumentation = !opt.qemu_mode
+        && !opt.frida_mode
+        && !opt.unicorn_mode
+        && !opt.forkserver_cs
+        && !opt.non_instrumented_mode
+        && !opt.nyx_mode;
+
+    if check_instrumentation && !is_instrumented(&mmap, shmem_env_var) {
+        return Err(Error::illegal_argument(
+            "target binary is not instrumented correctly",
+        ));
+    }
+
+    if opt.forkserver_cs || opt.qemu_mode || opt.frida_mode && is_instrumented(&mmap, shmem_env_var)
+    {
+        return Err(Error::illegal_argument("Instrumentation found in -Q mode"));
+    }
+
+    if mmap_has_substr(&mmap, "__asan_init")
+        || mmap_has_substr(&mmap, "__lsan_init")
+        || mmap_has_substr(&mmap, "__lsan_init")
+    {
+        opt.uses_asan = true;
+    }
+
+    if mmap_has_substr(&mmap, PERSIST_SIG) {
+        opt.is_persistent = true;
+    } else if opt.is_persistent {
+        println!("persistent mode enforced");
+    } else if opt.frida_persistent_addr.is_some() {
+        println!("FRIDA persistent mode configuration options detected");
+    }
+
+    if opt.frida_mode || mmap_has_substr(&mmap, DEFER_SIG) {
+        println!("deferred forkserver binary detected");
+        opt.defer_forkserver = true;
+    } else if opt.defer_forkserver {
+        println!("defer forkserver enforced");
+    }
+
+    Ok(())
+    // Safety: unmap() is called on Mmap object Drop
+}
+
+fn mmap_has_substr(mmap: &Mmap, sub_str: &str) -> bool {
+    let mmap_len = mmap.len();
+    let substr_len = sub_str.len();
+    if mmap_len < substr_len {
+        return false;
+    }
+    for i in 0..(mmap_len - substr_len) {
+        if &mmap[i..i + substr_len] == sub_str.as_bytes() {
+            return true;
+        }
+    }
+    false
+}
+
+fn is_instrumented(mmap: &Mmap, shmem_env_var: &str) -> bool {
+    mmap_has_substr(&mmap, shmem_env_var)
+}
+fn find_executable_in_path(executable: &Path) -> Option<PathBuf> {
```

Comment:
We use `which` a bunch, could be handy here as well(?)

Reply:
IMO this is fine, it's only 10 lines and we already have ``split_paths``

Reply:
well the crate does more than these 10 lines but sure